intel/geoip/database.go

```diff
@@ -13,106 +13,205 @@ import (
 }
 
-func doReload() error {
-	// reload if needed
-	if dbDoReload.SetToIf(true, false) {
-		closeDBs()
-		if err := openDBs(); err != nil {
-			// try again the next time
-			dbDoReload.SetTo(true)
-			return err
-		}
+// notifyWaiters notifies and removes all waiters. Must be called
+// with ub.rw locked.
+func (ub *updateBroadcaster) notifyWaiters() {
+	waiters := ub.waiters
+	ub.waiters = nil
+	for _, c := range waiters {
+		close(c)
 	}
+}
 
-	return nil
+// getWaiter appends and returns a new waiter channel that gets closed
+// when a new database version is available. Must be called with
+// ub.rw locked.
+func (ub *updateBroadcaster) getWaiter() chan struct{} {
+	ch := make(chan struct{})
+	ub.waiters = append(ub.waiters, ch)
+	return ch
 }
 
-func openDBs() error {
-	var err error
+type updateWorker struct {
+	trigger chan struct{}
+	once    sync.Once
 
-	geoDBv4File, err = updates.GetFile("intel/geoip/geoipv4.mmdb.gz")
-	if err != nil {
-		return fmt.Errorf("could not get GeoIP v4 database file: %s", err)
+	v4 updateBroadcaster
+	v6 updateBroadcaster
+}
+
+// GetReader returns a MMDB reader for either the IPv4 or the IPv6 database.
+// If wait is true GetReader will wait at most 1 second for the database to
+// become available. If no database is available or GetReader times-out while
+// waiting nil is returned.
+func (upd *updateWorker) GetReader(v6 bool, wait bool) *maxminddb.Reader {
+	// check which updateBroadcaster we need to use
+	var ub *updateBroadcaster = &upd.v4
+	if v6 {
+		ub = &upd.v6
 	}
-	unpackedV4, err := geoDBv4File.Unpack(".gz", updater.UnpackGZIP)
-	if err != nil {
-		return err
+
+	// lock the updateBroadcaster and - if we are allowed to wait -
+	// create a new waiter channel, trigger an update and wait for at
+	// least 1 second for the update to complete.
+	ub.rw.Lock()
+	if ub.db == nil {
+		if wait {
+			waiter := ub.getWaiter()
+			ub.rw.Unlock()
+
+			upd.triggerUpdate()
+
+			select {
+			case <-waiter:
+				// call this method again but this time we don't allow
+				// it to wait since there must be a open database anyway ...
+				return upd.GetReader(v6, false)
+			case <-time.After(time.Second):
+				// we tried hard but failed so give up here
+				return nil
+			}
+		}
+		ub.rw.Unlock()
+		return nil
 	}
-	geoDBv4Reader, err = maxminddb.Open(unpackedV4)
-	if err != nil {
-		return err
+	rd := ub.db.Reader
+	ub.rw.Unlock()
+
+	return rd
+}
+
+// triggerUpdate triggers a database update check.
+func (upd *updateWorker) triggerUpdate() {
+	upd.start()
+
+	select {
+	case upd.trigger <- struct{}{}:
+	default:
 	}
+}
 
-	geoDBv6File, err = updates.GetFile("intel/geoip/geoipv6.mmdb.gz")
-	if err != nil {
-		return fmt.Errorf("could not get GeoIP v6 database file: %s", err)
+func (upd *updateWorker) start() {
+	upd.once.Do(func() {
+		module.StartServiceWorker("geoip-updater", time.Second*10, upd.run)
+	})
+}
+
+func (upd *updateWorker) run(ctx context.Context) error {
+	firstTime := true
```

Comment:
Instead of using a "first" flag, you can just move to select block to the end of the for loop.

Reply:
yeah, I moved it to the top because I had two `continue` statements in early versions but they are not there anymore so I can move it to the bottom again. Thanks for pointing out